docs/rfcs/xxxx-integrations-in-operator.md

```diff
@@ -0,0 +1,265 @@
+---
+title: Integrations in Grafana Agent Operator
+weight: xxxx
+---
+
+# Integrations in Grafana Agent Operator
+
+* Date: 2022-01-04
+* Author: Robert Fratto (@rfratto)
+
+## Background
+
+Grafana Agent includes support for integrations, which are intended as
+"batteries-included" features to assist with collecting telemetry data. With
+the `integrations-next` feature enabled, there are multiple types of
+integrations:
+
+* Integrations that generate metrics (i.e., `node_exporter`)
+* Integrations that generate logs (i.e., `eventhandler`)
+* Integrations that generate other types of telemetry are planned (i.e., an
+  upcoming `app_o11y_receiver`)
+
+Generically, an integration is a specialized telemetry collector for some
+system under observation. For example, a `redis` integration collects telemetry
+for Redis. Integrations can generate any combination of Prometheus metrics,
+Grafana Loki logs, or Grafana Tempo traces.
+
+This document proposes adding a way to add support for all current and future
+integrations into the Grafana Agent Operator.
+
+This proposal supersedes [#883][], which was the first attempt at designing the
+feature. This proposal takes advantage of the lessons I've learned and
+minimizes the implementation effort.
+
+## Goals
+
+* Allow Grafana Agent Operator to deploy integrations
+* Allow deployed integrations to write telemetry data
+* Support integrations that must exist on every machine (i.e., `node_exporter`)
+* Minimize development effort for creating new integrations
+
+## Non-Goals
+
+* Support externally collecting metrics from integrations
+
+## Proposal
+
+At a high level, the proposal is to:
+
+* Define a new `Integration` CRD which specifies a single instance of an
+  integration to run.
+* Update `GrafanaAgent` to discover `Integration`s and run integrations.
+
+## Architecture
+
+### Running integrations
+
+The new CRD, Integration, will be used for supporting all current integrations.
+The spec of Integration primarily revolves around three fields:
+
+* `name`: The name of the integration (e.g., `node_exporter`, `mysqld_exporter`)
+* `type`: Information about the integration being deployed
+* `config`: YAML configuration block for the integration
+
+The `type` field is an object with the following fields:
+
+* `allNodes`: True when the `name` integration should run on all Kubernetes
+  Nodes.
+* `unique`: True when the `name` integration must be unique across a
```

Comment:
To motivate this would maybe provide a sample use case here / example of an integration where you'd need this to be the case

Reply:
Also, what happens if you say `allNodes: true` and `unique: true`. Not all combinations make sense. Maybe a single field with three possibilities is the way to go (although I don't love singleton to describe it either)

Reply:
Here's a truth table for relevant integrations:

| allNodes | unique | example integration |
| -------- | ------ | ------------------- |
| false    | false  | mysqld_exporter     |
| false    | true   | statsd_exporter     |
| true     | false  | process_exporter    |
| true     | true   | node_exporter       |

Some of the table is arbitrary; there's no reason you _couldn't_ deploy statsd_exporter on all nodes. 

Also it's noteworthy to say that allNodes has no bearing on the generated agent config, while unique does.